copy/progress_reader.go

```diff
@@ -9,20 +10,81 @@ import (
 
 // progressReader is a reader that reports its progress on an interval.
 type progressReader struct {
-	source   io.Reader
-	channel  chan types.ProgressProperties
-	interval time.Duration
-	artifact types.BlobInfo
-	lastTime time.Time
-	offset   uint64
+	source       io.Reader
+	channel      chan types.ProgressProperties
+	interval     time.Duration
+	artifact     types.BlobInfo
+	lastUpdate   time.Time
+	offset       uint64
+	offsetUpdate uint64
 }
 
+// newProgressReader creates a new progress reader for:
+// `channel`:  The reporter channel to which the progress will be sent
+// `interval`: The update interval to indicate how often the progress should update
+// `artifact`: The blob metadata which is currently being progressed
+func newProgressReader(
+	channel chan types.ProgressProperties,
+	interval time.Duration,
+	artifact types.BlobInfo,
+) *progressReader {
+	// The progress reader constructor informs the progress channel
+	// that a new artifact will be read
+	channel <- types.ProgressProperties{
+		Event:    types.ProgressEventNewArtifact,
+		Artifact: artifact,
+	}
+	return &progressReader{
+		source:       nil,
+		channel:      channel,
+		interval:     interval,
+		artifact:     artifact,
+		lastUpdate:   time.Now(),
+		offset:       0,
+		offsetUpdate: 0,
+	}
+}
+
+// withSource sets the provided source reader to the progressReader and creates
+// a new io.Reader from it
+func (r *progressReader) withSource(source io.Reader) io.Reader {
```

Comment:
“creates a new `io.Reader`” is incorrect.

I tend to read `withSource` as “make a copy with a changed source”.

If the caller calls `newProgressReader` and `withSource` immediately afterwards, there’s little benefit to splitting them — so if you don’t plan to move the initial step elsewhere, maybe just fold this into the constructor and we can avoid a naming discussion. (`progressReader` is package-private, so we can always split this out again if necessary.)

Reply:
Alright I will put it back into the `newProgressReader` function and remove this one. This should be now easier and more straight forward to use.